### ISSUE TITLE
Custom function, function of a variable

### ISSUE THREAD

**Issue by mgraffiedi** [sfepy/sfepy#1010]
Hi all,
I do not know how to write a material function that can take as input the variable T (temperature) in my case. Which is the unknown field variable.
Is that possible using Sfepy, and in case, how?
Best

Marco Graffiedi

**Comment by rc** [comment]
While this is possible (see [this example](http://sfepy.org/doc-devel/examples/diffusion-poisson_field_dependent_material.html)), it might be a better idea to use [this term](http://sfepy.org/doc-devel/src/sfepy/terms/terms_diffusion.html#sfepy.terms.terms_diffusion.NonlinearDiffusionTerm). There is no example at the moment, but check #998 - @peppe988 might have some input.

**Comment by peppe988** [comment]
Ciao Marco,
Here is an example to show you how to use the non-linear term @rc told you about, and I have also added another [term](http://sfepy.org/doc-devel/src/sfepy/terms/terms_volume.html#sfepy.terms.terms_volume.NonlinearVolumeForceTerm). You may ask now what is the difference between the 2 examples:
1) The one in the previous comment introduces a fictitious time dependence and at a single step the material function is seen as a constant, only updating it at the end.
2) My example takes into consideration the field-dependent function, so as you can see you need to supply the derivative.

Hope it helps, let me know if you have a problem with the code. it was actually on my to-do list to write an example, thanks for reminding me.
[sfepy example.zip](https://github.com/sfepy/sfepy/files/12134641/sfepy.example.zip)

**Comment by mgraffiedi** [comment]
Hi rc, hi Peppe
thanks!!!!! It worked great!

I was aware of the function dw_nl_diffusion, but now I understand how to use it.
Now, I understand that I would need the nonlinear version of BCNewtonTerm. However, I think that is not implemented yet, may I ask you to confirm this?

**Comment by peppe988** [comment]
yes, it seems so. But honestly, that term looks like a specific example of the dw_volume_nvf calculated on a surface. Maybe better to wait for @rc just to be sure. 
Anyhow if you look at https://github.com/sfepy/sfepy/issues/998 you will see that is not that complicated to write your own non linear terms, in case you need it.

**Comment by rc** [comment]
Yes, it looks like dw_volume_nvf might work with surface integrals - can you test it? The dw_bc_newton is just a shorthand for a dot term - a dot term with a known value.

**Comment by peppe988** [comment]
yes I have tested and it works @rc. @mgraffiedi I confirm that you just need to substitute the integration region with a surface, let me know if this solves your problem.
I have made these simple changes to the previous code:
[poisson_field_dependent_material.py.zip](https://github.com/sfepy/sfepy/files/12176657/poisson_field_dependent_material.py.zip)

**Comment by mgraffiedi** [comment]
Hi all!!!
this is good progress!!!! Thanks
However, there is something that I do not understand. the simulation gives me a constant temperature on the cylinder, regardless of the non-linear source term.

In my simulation, I have a constant source term on one surface and q(T) on another. So I tried to simulate with no ebcs, but in that case, it does not converge. 

So, I attach my case (simplified), which does not converge. If I impose a temperature at one edge, it will give me a reasonable solution that is independent of the q term (I was not expecting this)

[Sym_simplified.zip](https://github.com/sfepy/sfepy/files/12256973/Sym_simplified.zip)

**Comment by mgraffiedi** [comment]
No problem, there is no reason to apologize.

Regarding your code, I tried changing the nonlinear source in the cylinder problem ( Ciao to you ;) ). And the solution seems not to be changing. But, I may be doing something wrong here.

Regarding mine. Yes, in my simulation the two surfaces are superimposed. In the real world, one is a surface BC and the other is a volumetric heat generation that is concentrated very close to the surface. Considering how small the volume interested by heat generation is, I thought it would be better to simplify it as a surface source term
Best

**Comment by peppe988** [comment]
> No problem, there is no reason to apologize.
> 
> Regarding your code, I tried changing the nonlinear source in the cylinder problem ( Ciao to you ;) ). And the solution seems not to be changing. But, I may be doing something wrong here.

have you probed the solution with 'sfepy.postprocess.probes_vtk' to be sure that the changes are just not noticeable by eye? I do not know how much of an effect a non-linear term on the surface has on the solution of the equation.

> 
> Regarding mine. Yes, in my simulation the two surfaces are superimposed. In the real world, one is a surface BC and the other is a volumetric heat generation that is concentrated very close to the surface. Considering how small the volume interested by heat generation is, I thought it would be better to simplify it as a surface source term Best

ok. Anyway, try to use a better mesh just to be sure
